### PR TITLE
Add roswell.

### PR DESCRIPTION
**Pull request by vale981**
###### Motivation for this change
Although roswell isn't much use as a lisp implementation manager on nix, it nevertheless can be useful.

I have wrapped it, so that the correct lisp implementation is chosen.

It would be great if some people test it and report if it works.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 120763` at 07107bb6 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>roswell</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `python3`.
Near pkgs/development/roswell/default.nix:1:26:
```
  |
1 | { lib, stdenv, fetchgit, python3, coreutils, automake, gnumake, gnum4,
  |                          ^
```
</li>
    <li>warning: unused-argument

Unused argument: `openssl`.
Near pkgs/development/roswell/default.nix:2:52:
```
  |
2 | autoconf, curl, autoreconfHook, makeWrapper, sbcl, openssl}:
  |                                                    ^
```
</li>
    <li>warning: unused-argument

Unused argument: `coreutils`.
Near pkgs/development/roswell/default.nix:1:35:
```
  |
1 | { lib, stdenv, fetchgit, python3, coreutils, automake, gnumake, gnum4,
  |                                   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/roswell/default.nix:13:3:
```
   |
13 |   configurePhase = "sh bootstrap && ./configure --prefix $out";
   |   ^
```
</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -3991,6 +3991,16 @@
...
+  hiro98 = {
+    email = "hiro@protagon.space";
+    github = "vale981";
+    githubId = 4025991;
+    name = "Valentin Boettcher";
+    keys = [{
+      longkeyid = "rsa2048/0xC22D4DE4D7B32D19";
+      fingerprint = "45A9 9917 578C D629 9F5F  B5B4 C22D 4DE4 D7B3 2D19";
+    }];
+  };
```
Comment (legendofmiracles) [inline maintainers/maintainer-list.nix:4003]: Should  be a separate commit: `maintainers: add Valentin Boettcher`

File: pkgs/development/roswell/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  buildInputs = [ sbcl ];
+
+  meta = {
+    description = "Roswell is a Lisp implementation installer/manager, launcher, and much more!";
+    license = lib.licenses.mit;
+    maintainers = [lib.maintainers.hiro98];
+    platforms = lib.platforms.linux;
+    homepage = "https://github.com/roswell/roswell";
+  };
+}
```
Comment (legendofmiracles) [inline pkgs/development/roswell/default.nix:31]:
```suggestion
  meta = with lib; {
    description = "Roswell is a Lisp implementation installer/manager, launcher, and much more!";
    license = licenses.mit;
    maintainers = with maintainers; [ hiro98 ];
    platforms = platforms.linux;
    homepage = "https://github.com/roswell/roswell";
  };
}
```

File: pkgs/development/roswell/default.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+autoconf, curl, autoreconfHook, makeWrapper, sbcl}:
+
+stdenv.mkDerivation rec {
+  pname = "roswell";
+  version = "21.01.14.108";
+  src = fetchgit {
+    url = "https://github.com/roswell/roswell.git";
+    rev = "v${version}";
+    sha256 = "1hj9q3ig7naky3pb3jkl9yjc9xkg0k7js3glxicv0aqffx9hkp3p";
+  };
```
Comment (legendofmiracles) [inline pkgs/development/roswell/default.nix:11]:
```suggestion
{ lib, stdenv, fetchFromGitHub, automake, gnumake, gnum4,
autoconf, curl, autoreconfHook, makeWrapper, sbcl}:

stdenv.mkDerivation rec {
  pname = "roswell";
  version = "21.01.14.108";
  src = fetchFromGitHub {
    owner = "roswell";
    repo = pname;
    rev = "v${version}";
    sha256 = "1hj9q3ig7naky3pb3jkl9yjc9xkg0k7js3glxicv0aqffx9hkp3p";
  };
```

File: pkgs/development/roswell/default.nix [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  configurePhase = """
+    runHook preConfigure
+    sh bootstrap && ./configure --prefix $out
+    runHook postConfigure
+  """;
+
+  postFixup = "wrapProgram $out/bin/ros --add-flags 'lisp=sbcl-bin/system' --prefix PATH ${sbcl}/bin";
+
+  nativeBuildInputs = [ automake gnumake gnum4 autoconf curl autoreconfHook makeWrapper ];
```
Comment (legendofmiracles) [inline pkgs/development/roswell/default.nix:21]: IIRC automake and gnumake, and a few others should already be in stdenv

**Review by legendofmiracles — CHANGES_REQUESTED** [review]
Squash all package related commits please.

**Comment by vale981** [comment]
Thank you for the feedback. I've slimmed down the derivation and rebased everything.

File: pkgs/development/roswell/default.nix [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1hj9q3ig7naky3pb3jkl9yjc9xkg0k7js3glxicv0aqffx9hkp3p";
+  };
+
+  configurePhase = ''
+    runHook preConfigure
+    sh bootstrap && ./configure --prefix $out
+    runHook postConfigure
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/roswell/default.nix:17]:
```suggestion
  preConfigure = ''
    ./bootstrap
  '';
  
  configureFlags = [ "--prefix=$out" ];
```

File: pkgs/development/roswell/default.nix [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  postFixup = "wrapProgram $out/bin/ros --add-flags 'lisp=sbcl-bin/system sbcl-bin.version=system' --prefix PATH : ${sbcl}/bin --argv0 ros";
```
Comment (SuperSandro2000) [inline pkgs/development/roswell/default.nix:19]:
```suggestion
  postInstall = ''
    wrapProgram $out/bin/ros \
      --add-flags 'lisp=sbcl-bin/system sbcl-bin.version=system' \
      --prefix PATH : ${lib.makeBinPath [ sbcl ]} --argv0 ros
  '';
```

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -11496,6 +11496,8 @@ in
   sbcl_2_1_2 = callPackage ../development/compilers/sbcl/2.1.2.nix {};
   sbcl = sbcl_2_1_2;
 
+  roswell = callPackage ../development/roswell/default.nix {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:11499]:
```suggestion
  roswell = callPackage ../development/roswell/default.nix { };
```

Comment (vale981) [reply pkgs/top-level/all-packages.nix:11499]: why that?

File: pkgs/development/roswell/default.nix (hunk 5 above)
Comment (vale981) [reply pkgs/development/roswell/default.nix:17]: that breaks it bc. it tries to init setuptools "the wrong way"

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (legendofmiracles) [reply pkgs/top-level/all-packages.nix]: keeping things consistent.

File: pkgs/development/roswell/default.nix [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  nativeBuildInputs = [ autoreconfHook makeWrapper ];
+  buildInputs = [ sbcl curl ];
+
+  meta = with lib; {
+    description = "Roswell is a Lisp implementation installer/manager, launcher, and much more!";
+    license = licenses.mit;
+    maintainers = with maintainers; [ hiro98 ];
+    platforms = platforms.linux;
+    homepage = "https://github.com/roswell/roswell";
+  };
```
Comment (legendofmiracles) [inline pkgs/development/roswell/default.nix:34]:
```suggestion
    homepage = "https://github.com/roswell/roswell";
    mainProgram = "ros";
  };
```
Because the main binary is not the same name as the deriviation

File: pkgs/development/roswell/default.nix (hunk 5 above)
Comment (SuperSandro2000) [reply pkgs/development/roswell/default.nix:17]: We should use the configure phase by nix instead of overwriting everything.

Comment (vale981) [reply pkgs/development/roswell/default.nix:17]: have you tried building the derivation with that change?

File: pkgs/development/roswell/default.nix [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1hj9q3ig7naky3pb3jkl9yjc9xkg0k7js3glxicv0aqffx9hkp3p";
+  };
+
+  configurePhase = ''
+    runHook preConfigure
+    sh bootstrap && ./configure --prefix $out
+    runHook postConfigure
+  '';
```
Comment (OPNA2608) [inline pkgs/development/roswell/default.nix:17]:
```suggestion
  preConfigure = ''
    sh bootstrap
  '';

  configureFlags = [ "--prefix=${placeholder "out"}" ];
```

@SuperSandro2000 `$out` -> `${placeholder "out"}` in flags, otherwise you get garbage results.

> `/nix/store/dyikz835agazxvxm9fj7xqlsz0k34iwj-coreutils-8.32/bin/mkdir -p 'ut/bin'`

File: pkgs/development/roswell/default.nix [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    wrapProgram $out/bin/ros \
+      --add-flags 'lisp=sbcl-bin/system sbcl-bin.version=system' \
+      --prefix PATH : ${lib.makeBinPath [ sbcl ]} --argv0 ros
+  '';
+
+  nativeBuildInputs = [ autoreconfHook makeWrapper ];
+  buildInputs = [ sbcl curl ];
+
+  meta = with lib; {
+    description = "Roswell is a Lisp implementation installer/manager, launcher, and much more!";
```
Comment (OPNA2608) [inline pkgs/development/roswell/default.nix:29]: [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/e7e72e070b9a82730e97134bf33d6a7e608a6bf3/.github/CONTRIBUTING.md):

> `meta.description` should:
>   - […]
>   - Not have a period at the end.

It's an exclamation mark in your case but I'd extend that guideline to punctuation marks in general.

```suggestion
    description = "Roswell is a Lisp implementation installer/manager, launcher, and much more";
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Some suggestions, including a fixed version of https://github.com/NixOS/nixpkgs/pull/120763#discussion_r621051202.

> Have you tried building the derivation with that change?

Tested these changes, still builds for me.

File: pkgs/top-level/all-packages.nix [hunk 11]
```diff
@@ -11496,6 +11496,8 @@ in
   sbcl_2_1_2 = callPackage ../development/compilers/sbcl/2.1.2.nix {};
   sbcl = sbcl_2_1_2;
 
+  roswell = callPackage ../development/roswell/default.nix { };
```
Comment (OPNA2608) [inline pkgs/top-level/all-packages.nix:11499]: Please move this package to another location like `pkgs/development/tools/roswell/default.nix` and adjust this line accordingly, `pkgs/development/roswell` is too high up for a single package IMO.

**Comment by OPNA2608** [comment]
And please squash everything into 2 commits once you're done:

- `maintainers: add Valentin Boettcher`
- `roswell: init at 21.01.14.108`

File: pkgs/development/roswell/default.nix (hunk 9 above)
Comment (vale981) [reply pkgs/development/roswell/default.nix:17]: that must have been it :]
thanks

**Comment by vale981** [comment]
> And please squash everything into 2 commits once you're done:
> 
>     * `maintainers: add Valentin Boettcher`
> 
>     * `roswell: init at 21.01.14.108`

done, and thank you very much for reviewing and improving :]

File: pkgs/development/tools/roswell/default.nix [hunk 12]
```diff
@@ -0,0 +1,38 @@
+{ lib, stdenv, fetchFromGitHub, curl, autoconf, automake, makeWrapper, sbcl}:
```
Comment (OPNA2608) [inline pkgs/development/tools/roswell/default.nix:1]: I missed this, apologies.

```suggestion
{ lib, stdenv, fetchFromGitHub, curl, autoconf, automake, makeWrapper, sbcl }:
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 120763` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>roswell</li>
  </ul>
</details>

Binary seems to work, though I don't know how to use it. If you fix the styling nitpick and the tool works fine for you then LGTM.

**Review by OPNA2608 — APPROVED** [review]

**Comment by vale981** [comment]
@OPNA2608  Sorry, I've accidentally re-requested a review from you.

**Review by legendofmiracles — APPROVED** [review]
